### PR TITLE
Homework - modern cpp

### PR DESCRIPTION
**Pull request by arafalski**

### CODE REVIEW

File: shapes/main.cpp [hunk 1]
```diff
@@ -1,78 +1,88 @@
...
-                                     bool (*predicate)(shared_ptr<Shape> s),
-                                     std::string info)
-{
-    Collection::const_iterator iter = std::find_if(collection.begin(), collection.end(), predicate);
-    if(*iter != nullptr)
-    {
+                                     std::function<bool(shared_ptr<Shape> s)> predicate,
+                                     std::string info) {
+    auto iter = std::find_if(collection.begin(), collection.end(), predicate);
+    if (*iter != nullptr) {
```
Comment (WBigaj) [inline shapes/main.cpp]:
```suggestion
    if (*iter) {
```

File: shapes/main.cpp [hunk 2]
```diff
@@ -1,78 +1,88 @@
...
 
-void printAreas(const Collection& collection)
-{
-    for(vector<shared_ptr<Shape>>::const_iterator it = collection.begin(); it != collection.end(); ++it)
-        if(*it)
-            cout << (*it)->getArea() << std::endl;
+void printAreas(const Collection& collection) {
+    for (const auto& it : collection)
+        if (it)
+            cout << (it)->getArea() << std::endl;
```
Comment (WBigaj) [inline shapes/main.cpp]:
```suggestion
        if (it) {
            cout << (it)->getArea() << std::endl;
        }
```

File: shapes/main.cpp [hunk 3]
```diff
@@ -1,78 +1,88 @@
...
 
-void printCollectionElements(const Collection& collection)
-{
-    for(Collection::const_iterator it = collection.begin(); it != collection.end(); ++it)
-        if(*it)
-            (*it)->print();
+void printCollectionElements(const Collection& collection) {
+    for (const auto& it : collection)
+        if (it)
+            printShape(*it);
```
Comment (WBigaj) [inline shapes/main.cpp]:
```suggestion
        if (it) {
            printShape(*it);
        }
```

File: shapes/main.cpp [hunk 4]
```diff
@@ -1,78 +1,88 @@
...
 
-bool areaLessThan10(shared_ptr<Shape> s)
-{
-    if(s)
-        return (s->getArea() < 10);
+std::function<bool(shared_ptr<Shape>)> areaLessThanX{[x{10}](shared_ptr<Shape> s) {
+    if (s)
+        return (s->getArea() < x);
     return false;
+}};
```
Comment (WBigaj) [inline shapes/main.cpp]:
```suggestion
    return s ? s->getArea() < x : false;
}};
```

**Review by WBigaj — COMMENTED** [review]
I have checked your code and left my feedback.

File: shapes/main.cpp (hunk 4 above)
Comment (MaciejSalamonski) [reply shapes/main.cpp]: +1

**Review by MaciejSalamonski — APPROVED** [review]
Nice code 👍

**Review by jakjan95 — APPROVED** [review]
For me code looks good, just some {} are missing, but that was not a homework task.

File: shapes/Circle.hpp [hunk 5]
```diff
@@ -2,19 +2,21 @@
...
 #include "Shape.hpp"
 
-class Circle : public Shape
-{
+class alignas(128) Circle final : public Shape {
 public:
-    Circle(double r);
-    Circle(const Circle & other);
+    using Shape::Shape;
+    Circle(double r, Color color = Color::Green);
```
Comment (ziobron) [inline shapes/Circle.hpp]: I do not see delegating constructors here. The idea was to leave the old constructor and call the new one in the olds' initialization list:

```suggestion
    Circle(double r) : Circle(r, Color::Green) {}
    Circle(double r, Color color);
```

calling constructor from it's own class was not allowed before C++14.
The same for other classes.

PS. I know that your version is more intuitive and does the same, but idea here is to learn how does delegating constructor works.

**Review by ziobron — CHANGES_REQUESTED** [review]
Only one this is missing - delegating constructors

**Review by ziobron — APPROVED** [review]

**Comment by ziobron** [comment]
15 XP za Nowoczesny C++ SFINAE (4447)
5 XP za Nowoczesny C++ alignas (4448)
5 XP za Nowoczesny C++ delegating constructors (4449)
5 XP za Nowoczesny C++ lambda (4450)
5 XP za Nowoczesny C++ structured bindings (4451)
15 XP za Nowoczesny C++ variadic templates (4452)

{ "requirement_ids": [4447, 4448, 4449, 4450, 4451, 4452] }

**Pull request closed without merging**